Add the ability to manage the entry types from the Solidus admin panel

vassalloandrea opened the pull request:
Ref: #14

vassalloandrea committed:
Add the Content link to the Solidus admin panel

vassalloandrea committed:
Show the list of entry types paginated

vassalloandrea committed:
Add the ability to delete an entry type from the admin

vassalloandrea committed:
Add the ability to create an entry type

elia committed:
Add an admin-hint for new entry-types

Co-authored-by: Andrea Vassallo <andrea.vassallo.94@gmail.com>

elia committed:
Delegate provider names translations to SolidusContent

Co-authored-by: Andrea Vassallo <andrea.vassallo.94@gmail.com>

elia committed:
Rely on default resource-controller and rails helper

Co-authored-by: Andrea Vassallo <andrea.vassallo.94@gmail.com>

elia committed:
Add admin/entry_types/edit

It uses the form as it is, we'll add the ability to add options later.

Co-authored-by: Andrea Vassallo <andrea.vassallo.94@gmail.com>

elia committed:
Restore default resource-controller url helpers

Co-authored-by: Andrea Vassallo <andrea.vassallo.94@gmail.com>

elia committed:
Extract common resource controller fixes to an abstract class

elia committed:
Add an inline hint for provider name

elia committed:
Prevent provider from being changed after creation

elia committed:
Accept the deletion confirmation modal in spec

elia committed:
Whitespace

elia committed:
Allow insertion of EntryTypes options

This is done through a text-area as an MVP, will be improved using a
full-blown form later.

Co-authored-by: Andrea Vassallo <andrea.vassallo.94@gmail.com>

elia committed:
Merge all entry type feature specs in a single file

aldesantis commented on app/models/solidus_content/entry_type.rb:
```diff
@@ -19,4 +26,32 @@ def content_for(entry)
+    persisted?
+  end
+
+  def options=value
```
Missing parentheses?

aldesantis commented on app/models/solidus_content/entry_type.rb:
```diff
@@ -19,4 +26,32 @@ def content_for(entry)
+      begin
+        value = JSON.parse(value)
+      rescue JSON::ParserError
+        value
```
Should we swallow errors like this? What's the use case for non-JSON content in the `options` attribute?

elia replied:
You're right, there is an underlying problem that is: how to accept with serialized  json? where should we handle it before feeding it to the model?

I'm now updating the code to use the ActiveRecord Json type to do the serialization, [it still has a `rescue nil`](https://github.com/rails/rails/blob/6-0-stable/activerecord/lib/active_record/type/json.rb#L14) in it, but it's library code and no one cares about that 😄, jk, library code is the place in which to do this kind of stuff 👍

aldesantis reviewed (changes requested):
Very nice, just a couple of questions!

aldesantis commented on app/controllers/solidus_content/resource_controller.rb:
```diff
@@ -0,0 +1,27 @@
+      "SolidusContent::#{controller_name.classify}".constantize
+    end
+
+    def new_object_url(options = {})
+      new_admin_entry_type_path(options)
+    end
+
+    def edit_object_url(object, options = {})
+      edit_admin_entry_type_path(object, options)
+    end
+
+    def object_url(object = nil, options = {})
+      admin_entry_type_path((object || @object), options)
+    end
+
+    def collection_url(options = {})
+      admin_entry_types_path(options)
+    end
```
Why do you need to override these?

vassalloandrea replied:
This is needed since the `Spree::ResourceController` uses the `Spree` namespace to build the resource URL and path helpers.

This extension uses the SolidusContent namespace instead.

vassalloandrea committed:
Fix create new entry_type  options field null validation error

https://github.com/solidusio/solidus/blob/3724171102e737afa6afed5f8eb9ed58c0063e7b/backend/app/controllers/spree/admin/resource_controller.rb#L57
